### DISCUSSION TITLE
Adding Registries

### DISCUSSION THREAD

**Discussion by CMorrison82z** [mason-org/mason.nvim#1369] [Q&A] [answered]
Preface : I am new to the vim, linux ecosystem, so apologies for the ignorance on my part.

tl;dr : Please provide a sample for adding 'lua' registry type (and 'github' type for that matter. I was unable to add my fork). Additionally, consider 'json' registry type ?

I've added my own lua registry, using a single package from the mason-registry as a test. I took and converted a package entry from the .json from the releases page, converted to a lua table, then add it as the only entry in the returned registry (a list of packages).

I source it using absolute path to the "registry.lua". Mason recognizes it, but fails to install the test package I included.

Source :

```lua
return {
  {
    name = "lua-language-server",
    description = "A language server that offers Lua language support - programmed in Lua.",
    homepage = "https://github.com/LuaLS/lua-language-server",
    licenses = {
      "MIT"
    },
    languages = {
      "Lua"
    },
    categories = {
      "LSP"
    },
    source = {
      id = "pkg:github/LuaLS/lua-language-server@3.6.22",
      asset = {
        {
          target = "darwin_arm64",
          file = "lua-language-server-{{version}}-darwin-arm64.tar.gz:libexec/",
          bin = "exec:libexec/bin/lua-language-server"
        },
        {
          target = "darwin_x64",
          file = "lua-language-server-{{version}}-darwin-x64.tar.gz:libexec/",
          bin = "exec:libexec/bin/lua-language-server"
        },
        {
          target = "linux_arm64_gnu",
          file = "lua-language-server-{{version}}-linux-arm64.tar.gz:libexec/",
          bin = "exec:libexec/bin/lua-language-server"
        },
        {
          target = "linux_x64_gnu",
          file = "lua-language-server-{{version}}-linux-x64.tar.gz:libexec/",
          bin = "exec:libexec/bin/lua-language-server"
        },
        {
          target = "linux_x64_musl",
          file = "lua-language-server-{{version}}-linux-x64-musl.tar.gz:libexec/",
          bin = "exec:libexec/bin/lua-language-server"
        },
        {
          target = "win_x86",
          file = "lua-language-server-{{version}}-win32-ia32.zip",
          bin = "bin/lua-language-server.exe"
        },
        {
          target = "win_x64",
          file = "lua-language-server-{{version}}-win32-x64.zip",
          bin = "bin/lua-language-server.exe"
        }
      }
    },
    schemas = {
      lsp = "vscode:https://raw.githubusercontent.com/LuaLS/vscode-lua/master/package.json"
    },
    bin = {
      ["lua-language-server"] = "{{source.asset.bin}}"
    }
  }
}
```

**Comment by williamboman** [comment, marked as answer]
Hello! Custom registries is half-supported in the sense that while the settings allows you to register custom registries, many of the APIs involved in making a registry is not really made stable & public yet, hence no docs for custom registries yet.

Mason currently supports two types of registry sources, `github` and `lua`. `github` is the newer one and is used for https://github.com/mason-org/mason-registry. Creating a custom `github` registry is currently a bit tedious as there are no tools for this, you'll essentially have to manually recreate the setup in place at https://github.com/mason-org/mason-registry. I plan to make this easier in the form of reusable GitHub actions and a repo template soon :tm:.

`lua` is the legacy registry source that I'm not yet sure what to do with going forward (either drop support entirely or make some core changes). To use a `lua` registry source you'll need:

1) An index module. This is a Lua module (that can be `require()`'d by Neovim) returning a single table containing all available packages:

```lua
-- ~/.config/nvim/lua/custom-registry/init.lua
return {
    ["lua-linter"] = "custom-registry.packages.lua-linter",
    ["rust-linter"] = "custom-registry.packages.rust-linter",
}
```

2) For each package in the index, a Lua module returning a `Package` instance. `Pkg.new()` expects a table conforming to either:

* [`PackageSpec`](https://github.com/williamboman/mason.nvim/blob/main/doc/reference.md#packagespec) - This allows you to provide an entirely custom Lua function as the installer, or
* [`RegistryPackageSpec`](https://github.com/williamboman/mason.nvim/blob/main/doc/reference.md#registrypackagespec) - This is an entirely static data table describing how the package can be installed. [Here's the full table schema](https://github.com/mason-org/mason-registry/blob/main/CONTRIBUTING.md#package-specification). `mason.nvim` will parse this table and then compile an installation function when you install the package. This is what you're currently using.

In the example below I'll be using the `PackageSpec` variant (note: this variant is now legacy and not used by any core registries provided by Mason, and a package using this variant may break in the future):

```lua
-- ~/.config/nvim/lua/custom-registry/packages/lua-linter.lua
local Pkg = require "mason-core.package"

return Pkg.new {
  name = "lua-linter",
  desc = "Lua Linter.",
  homepage = "https://lua-linter.org",
  categories = { Pkg.Cat.Linter },
  languages = { Pkg.Lang.Lua },
  install = function (ctx)
     -- ctx: https://github.com/williamboman/mason.nvim/blob/main/doc/reference.md#installcontext
     ctx.spawn.luarocks { "install", "lua-linter" }
     ctx.fs:write_file("some-file", "some file contents idk")
     -- …

     ctx.receipt:with_primary_source {
        type = "unmanaged"
     }
  end
}
```

3) To finally use the registry you need to tell mason.nvim where to find it:

```lua
require("mason").setup {
  registries = {
    "lua:custom-registry", -- "custom-registry" here is what you'd pass to require() the index module (see 1) above)
    "github:mason-org/mason-registry",
  }
}
```

**Reply by williamboman** [reply]
And again as a final note, I'm currently 50/50 whether support for Lua-based packages should even remain. Either it'll be dropped entirely, or some additional requirements will be added to enable better package metadata management.

**Reply by CMorrison82z** [reply]
Well I'd really like to see some sort of support for a local registry (I know the packages therein are still remote, but still, I think the fewer remote dependencies, the better).

Thanks for clarifying, I'll have to give this a try later.

**Reply by acid-bong** [reply]
Just tried doing it this way (but changed npm to pypi -- I want Mason to install pyright-python wrapper with pip), but it seems to ignore my new registry.

Neovim: 0.9.0
Mason: upgraded today
Registry init: https://bpa.st/R7LNW
~~Example package: https://bpa.st/5DT26~~

Should I open an issue, if there's no similar ones?

Edit: shoulda edit it like this [commit](https://github.com/baco/mason.nvim/commit/f8cd7defd144194a714ba75aa600318158d82359) by baco
